### ISSUE TITLE
Kubemark-scale failing due to couple of hollow-nodes dead

### ISSUE THREAD

**Issue by shyamjvs** [kubernetes/kubernetes#60035]
We're observing almost continuous failures of our [kubemark-5000 job](https://k8s-testgrid.appspot.com/sig-scalability-kubemark#kubemark-5000):
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-kubemark-gce-scale/864
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-kubemark-gce-scale/865
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-kubemark-gce-scale/866
- https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-kubemark-gce-scale/867

And all of them are more or less resulting from a couple of hollow nodes getting unready:

```
I0218 16:42:35.334] Feb 18 16:42:35.333: INFO: Condition Ready of node hollow-node-2zgpj is false instead of true. Reason: NodeStatusUnknown, message: Kubelet stopped posting node status.
I0218 16:42:35.335] Feb 18 16:42:35.335: INFO: Condition Ready of node hollow-node-s64qb is false instead of true. Reason: NodeStatusUnknown, message: Kubelet stopped posting node status.
I0218 16:42:35.335] Feb 18 16:42:35.335: INFO: All nodes should be ready after test, Not ready nodes: ", hollow-node-2zgpj, hollow-node-s64qb"
```

These hollow nodes getting unready are causing failure by:
- either failing to start next test (due to failing `BeforeEach`)
- or timing out while trying to delete pods from those hollow-nodes (as their kubelets are gone)

I'll take a look into this further.
/assign

cc @kubernetes/sig-scalability-bugs @wojtek-t

**Comment by shyamjvs** [comment]
I looked into the currently running kubemark cluster and it seems that those hollow-nodes are being preempted to admit some critical pod:

```
Name:           hollow-node-zgmt5
Namespace:      kubemark
Node:           kubemark-5000-minion-group-rkj7/
Start Time:     Mon, 19 Feb 2018 09:26:36 +0000
Labels:         name=hollow-node
Annotations:    <none>
Status:         Failed
Reason:         Preempting
Message:        Preempted in order to admit critical pod
```

**Comment by shyamjvs** [comment]
This seems to be the case from the controller-manger logs:

```
I0219 09:40:24.932867       1 daemon_controller.go:912] Nodes needing daemon pods for daemon set fluentd-gcp-v2.0.16: [kubemark-5000-minion-group-rkj7], creating 1
```

**Comment by shyamjvs** [comment]
I'm looking at the kubelet logs on that node `kubemark-5000-minion-group-rkj7` and that gives me the final confirmation that fluentd is indeed the one behind it:

```
I0219 09:40:24.891334    1737 config.go:405] Receiving a new pod "fluentd-gcp-v2.0.16-6lqks_kube-system(e93f90b3-1558-11e8-9f1b-42010a8e0002)"
I0219 09:40:24.892659    1737 kubelet.go:1816] SyncLoop (ADD, "api"): "fluentd-gcp-v2.0.16-6lqks_kube-system(e93f90b3-1558-11e8-9f1b-42010a8e0002)"
... in order to free up resources: [(res: memory, q: 277442304), ]
I0219 09:40:24.893380    1737 kuberuntime_container.go:578] Killing container "docker://46ef5d08ccefc2383842b48af0d8fa6d52cb56b2817c0751c03153b0b3754e56" with 30 second grace period
I0219 09:40:24.893397    1737 kuberuntime_container.go:578] Killing container "docker://b3d0292480110e1079b150a8bef4967d2e700f450ad225445d0dc50caeae2e3d" with 30 second grace period
I0219 09:40:24.893488    1737 kuberuntime_container.go:578] Killing container "docker://a00ce7bff5b9ba28c17452934359f41d03ade5cac366f3209a307b357db1b948" with 30 second grace period
I0219 09:40:24.894639    1737 server.go:424] Event(v1.ObjectReference{Kind:"Pod", Namespace:"kubemark", Name:"hollow-node-zgmt5", UID:"fb8cfd1e-1556-11e8-9f1b-42010a8e0002", APIVersion:"v1", ResourceVersion:"24321", FieldPath:""}): type: 'Warning' reason: 'Preempting' Preempted in order to admit critical pod
```

**Comment by shyamjvs** [comment]
So it seems that the daemonset-controller has not yet created the pod for a node, while the hollow-nodes already got scheduled and started running there. Later when the pod comes, some hollow-node is being evicted to make space. I checked that the cpu/mem requests/limits for fluentd have stayed the same. There was a recent change https://github.com/kubernetes/kubernetes/pull/59657 around enabling fluentd resources scaling, but IIUC that is pretty much a no-op wrt to defaults (@x13n can confirm). Also from the fluentd pods, the requests (100m, 200MB) and limits (300 MB) seem to be the same as before

Why this was not happening before then?

My strong belief is that we've just started crossing the threshold after my recent [fix](https://github.com/kubernetes/kubernetes/pull/59832) around endpoints in kubemark. So hollow-proxies must be seeing considerable increase in mem usage. We were already quite close to the capacity earlier iiuc per holl:
- 80 n1-standard-8 VMs (each 30GB)
- 5k hollow-nodes = ~63 per real node
- 100M (kubelet) + ~350M (kubeproxy) + 20M (npd) = 470M per hollow-node

=> ~29 GB per node used by hollow-nodes

**Comment by shyamjvs** [comment]
Why did we see 1 green run after my endpoints fix (run 862) then?

It's a hypothesis, but I believe that we just got lucky with that run there due to hollow-nodes being distributed quite "equitably" across real nodes that we managed to avoid any pre-emption. IMO we should bump our real nodes by 1 and check if that works. Btw - note that we're seeing just 0-2 hollownode preemptions right now (i.e borderline).

**Comment by x13n** [comment]
There may be a small change in behavior after https://github.com/kubernetes/kubernetes/pull/59657: fluentd-gcp will be created without any resource request and then the request will be set after a short while. This probably didn't cause the problem you've seen, but I thought may be worth noting anyway.

**Comment by shyamjvs** [comment]
Yeah.. that doesn't seem to be the case here (from the logs). Thanks for the heads up @x13n.

**Comment by shyamjvs** [comment]
I'm reopening this in light of https://github.com/kubernetes/kubernetes/issues/61189#issuecomment-373641543.
The real reason why this might've happened maybe fluentd-scaler (noted in https://github.com/kubernetes/kubernetes/issues/60035#issuecomment-366684544) instead of increased mem-usage. Needs verification.
